Adx stake rebased

Adrena-Corto opened the pull request:
Follow up of the [first branch](https://github.com/AdrenaDEX/perpetuals/pull/5) 

_____

- [x] add_stake IX
  - [x] add a call to claim
- [x] remove_stake IX
  - [x] add a call to claim
- [x] claim_stake IX
  - [x] add decentralised claim incentive after cutoff period
- [x] resolve_stake_round IX
- [x] add a mechanism that convert collected protocol fee to the target mint and make them collectable in the protocol fees
  - [x] open position 
  - [x] close position 
  - [x] swap position
  - [x]  remove liquidity
  - [x] add liquidity

- [x] add voting power to staker similarly to vester

- [x] governance now use a shadow token

- [ ] add views for whatever front end need/might need later? (or add later)

Adrena-Corto committed:
LM token fees rebates incentives

acamill commented on programs/perpetuals/src/state/cortex.rs:
```diff
@@ -0,0 +1,119 @@
+
+        math::checked_div(
+            Self::INCEPTION_EMISSION_RATE as u64,
+            std::cmp::max(elapsed_epoches / 10, 1) as u64,
```
Yep, let's call this an ADRENA_EPOCH?

acamill commented on programs/perpetuals/src/state/cortex.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    fn get_emission_rate(inception_epoch: u64, current_epoch: u64) -> Result<u64> {
+        let elapsed_epoches = std::cmp::max(math::checked_sub(current_epoch, inception_epoch)?, 1);
```
Done

Adrena-Corto commented on programs/perpetuals/src/instructions/add_collateral.rs:
```diff
@@ -134,6 +154,20 @@ pub fn add_collateral(ctx: Context<AddCollateral>, params: &AddCollateralParams)
 
+    // compute amount of lm token to mint
+    let cortex = ctx.accounts.cortex.as_mut();
+    let lm_rewards_amount = cortex.get_lm_rewards_amount(fee_amount)?;
```
Fixed ty!

Adrena-Corto committed:
TMP (back to fixing tests on adx)

Adrena-Corto committed:
Add tests (not working)

Commit:
Start Spl governance integration

Adrena-Corto committed:
Add vest working + test (missing govpart)
Claim vest with half test (missing gov part)

Commit:
Setup vote with vestng

Adrena-Corto committed:
Update init IX

Adrena-Corto committed:
add_stake IX

Adrena-Corto committed:
- update add_stake ix

Adrena-Corto committed:
Remove_stake IX

Adrena-Corto committed:
WIP not working CPI to claim_stake

Adrena-Corto committed:
Working add/remove stake

Adrena-Corto committed:
Claim stake tests

Adrena-Corto committed:
resolved_staking_round

Adrena-Corto committed:
PR Feedback

Adrena-Corto committed:
Update basic test suite

Adrena-Corto committed:
remove artefact

Adrena-Corto committed:
WIP tests for staking

Adrena-Corto committed:
builds after rebase + fmt fixes

Adrena-Corto committed:
Tweaks

Adrena-Corto committed:
Staking
- convert collected fees to staking reward mint and make it available (only on open position and swap for now)
- stable, basic test working

the-orex reviewed (commented):
1st batch of comments

Adrena-Corto committed:
remove comment

Adrena-Corto committed:
Before switching to NON delegation with membership

Adrena-Corto committed:
Without delegation

Adrena-Corto committed:
Merge pull request #6 from AdrenaDEX/shadow-gov-token

Shadow gov token. Will be reviewed in the other PR. Not fully working missing a call with governance team to unlock depositing from non owner

Adrena-Corto committed:
Signer seed

Adrena-Corto committed:
PR FEEDBACK - Duplicate accounting

Adrena-Corto committed:
PR FEEDBACK - Remove useless comment

Adrena-Corto committed:
PD FEEDBACK - Improve comment

Adrena-Corto committed:
PR FEEDBACK - Remove empty account preprocessor

Adrena-Corto committed:
PR FEEDBACK - Add braces around fee distrib code in add liquidity

Adrena-Corto committed:
Ditto

Adrena-Corto committed:
PR FEEDBACK - Move relevant code in it's context

Adrena-Corto committed:
PR FEEDBACK - Scope the LM rewards

Adrena-Corto committed:
PR FEEDBACK - Scope claim

Adrena-Corto committed:
PR Feedback - remove duplicates accounting

Adrena-Corto committed:
Update programs/perpetuals/src/state/stake.rs

Co-authored-by: Orex <122088118+adrena-orex@users.noreply.github.com>

Adrena-Corto committed:
Update programs/perpetuals/src/state/custody.rs

Co-authored-by: Orex <122088118+adrena-orex@users.noreply.github.com>

Adrena-Corto committed:
Update programs/perpetuals/src/state/cortex.rs

Co-authored-by: Orex <122088118+adrena-orex@users.noreply.github.com>

Adrena-Corto committed:
Update programs/perpetuals/src/state/cortex.rs

Co-authored-by: Orex <122088118+adrena-orex@users.noreply.github.com>

Adrena-Corto committed:
Working tests

Adrena-Corto committed:
Fix custody save order (fix test)

Adrena-Corto committed:
All test working except liquidate and max_user_profit
